core/network-libp2p/src/service_task.rs

```diff
@@ -221,10 +221,71 @@ struct NodeInfo {
 	endpoint: ConnectedPoint,
 	/// Version reported by the remote, or `None` if unknown.
 	client_version: Option<String>,
+	/// Latest ping time with this node.
+	latest_ping: Option<Duration>,
 }
 
 impl<TMessage> Service<TMessage>
 where TMessage: CustomMessage + Send + 'static {
+	/// Returns a struct containing tons of useful information about the network.
+	pub fn state(&mut self) -> NetworkState {
+		let now = Instant::now();
+
+		let connected_peers = {
+			let swarm = &mut self.swarm;
+			self.nodes_info.values().map(move |info| {
+				let known_addresses = swarm.known_addresses(&info.peer_id)
+					.map(|(a, s)| (a.clone(), s)).collect();
```

Comment:
Why not `cloned()` instead of the map?

Reply:
The `(a, s)` is of type `(&Multiaddr, u32)`. `cloned()` would clone the tuple but not the multiaddr.

Reply:
Ahh yeah, I'm always failing with `cloned` :(